internal/clients/http/azure/create_vm.go

```diff
@@ -0,0 +1,430 @@
+						SourcePortRange:          to.Ptr("*"),
+						DestinationAddressPrefix: to.Ptr("0.0.0.0/0"),
+						DestinationPortRange:     to.Ptr("22"),
+						Protocol:                 to.Ptr(armnetwork.SecurityRuleProtocolTCP),
+						Access:                   to.Ptr(armnetwork.SecurityRuleAccessAllow),
+						Priority:                 to.Ptr[int32](100),
+						Description:              to.Ptr("network security group outbound port 22"),
+						Direction:                to.Ptr(armnetwork.SecurityRuleDirectionOutbound),
+					},
+				},
+			},
+		},
+	}
+
+	pollerResponse, err := nsgClient.BeginCreateOrUpdate(ctx, resourceGroupName, name, parameters, nil)
+	if err != nil {
+		return nil, fmt.Errorf("create of network security group failed to start: %w", err)
+	}
+
+	resp, err := pollerResponse.PollUntilDone(ctx, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to poll for create network security group result: %w", err)
+	}
+	return &resp.SecurityGroup, nil
+}
+
+func (c *client) createNetworkInterface(ctx context.Context, location string, resourceGroupName string, subnet *armnetwork.Subnet, publicIP *armnetwork.PublicIPAddress, nsg *armnetwork.SecurityGroup, name string) (*armnetwork.Interface, error) {
+	ctx, span := otel.Tracer(TraceName).Start(ctx, "createNetworkInterface")
+	defer span.End()
+
+	nicClient, err := c.newInterfacesClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	parameters := armnetwork.Interface{
+		Location: to.Ptr(location),
+		Properties: &armnetwork.InterfacePropertiesFormat{
+			IPConfigurations: []*armnetwork.InterfaceIPConfiguration{
+				{
+					Name: to.Ptr("ipConfig"),
+					Properties: &armnetwork.InterfaceIPConfigurationPropertiesFormat{
+						PrivateIPAllocationMethod: to.Ptr(armnetwork.IPAllocationMethodDynamic),
+						Subnet: &armnetwork.Subnet{
+							ID: subnet.ID,
+						},
+						PublicIPAddress: &armnetwork.PublicIPAddress{
+							ID: publicIP.ID,
+						},
+					},
+				},
+			},
+			NetworkSecurityGroup: &armnetwork.SecurityGroup{
+				ID: nsg.ID,
+			},
+		},
+	}
+
+	pollerResponse, err := nicClient.BeginCreateOrUpdate(ctx, resourceGroupName, name, parameters, nil)
+	if err != nil {
+		return nil, fmt.Errorf("create of network interface failed to start: %w", err)
+	}
+
+	resp, err := pollerResponse.PollUntilDone(ctx, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to poll for create network interface result: %w", err)
+	}
+
+	return &resp.Interface, nil
+}
+
+func (c *client) createVirtualMachine(ctx context.Context, location string, resourceGroupName string, instanceType armcompute.VirtualMachineSizeTypes, networkInterface *armnetwork.Interface, imageID string, sshKeyBody string, vmName string, diskName string) (*armcompute.VirtualMachine, error) {
+	ctx, span := otel.Tracer(TraceName).Start(ctx, "createVirtualMachine")
+	defer span.End()
+
+	vmClient, err := c.newVirtualMachinesClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	parameters := armcompute.VirtualMachine{
```

Comment:
This could be split so that the parameters variable is set before calling createVirtualMachine. We would get rid of few parameters and it would be a bit more readable.
Tbh I do not now how to decompose this in any other way. We could call
```
parameters := getParameters(location, armcompute.VirtualMachineSizeTypes(instanceType), networkInterface, imageID, pubkey.Body, vmName, diskName)
virtualMachine, err := c.createVirtualMachine(ctx, resourceGroupName, vmName, parameters)
```